spec/p2p/reactor/reactor.qnt

```diff
@@ -0,0 +1,233 @@
+/*
+ * Reactor is responsible for handling incoming messages on one or more
+ * Channel. Switch calls GetChannels when reactor is added to it. When a new
+ * peer joins our node, InitPeer and AddPeer are called. RemovePeer is called
+ * when the peer is stopped. Receive is called when a message is received on a
+ * channel associated with this reactor.
+ */
+// Code: https://github.com/cometbft/cometbft/blob/main/p2p/base_reactor.go
+module reactor {
+
+	// Unique ID of a node.
+	type NodeID = str
+
+	/*
+	 * Peer is an interface representing a peer connected on a reactor.
+	 */
+	type Peer = {
+		ID: NodeID,
+
+		// Other fields can be added to represent the p2p operation.
+	}
+
+	// Byte ID used by channels.
+	type Byte = int
+
+	// Channel configuration.
+	// Only the Byte ID is relevant here.
+	type ChannelDescriptor = {
+		ID: Byte,
+		Priority: int,
+	}
+
+	/*
+	 * Envelope contains a message with sender routing info.
+	 */
+	type Envelope = {
+		Src: Peer, // Sender
+		Message: str, // Payload
+		ChannelID: Byte,
+	}
+
+	// A Routine is used to interact with an active Peer.
+	type Routine = {
+		name: str,
+		peer: Peer,
+	}
+
+	type ReactorState = {
+		// Peers that have been initialized but not yet removed.
+		// The reactor should expect receiving messages from them.
+		peers: Set[NodeID],
+
+		// The reactor runs multiple routines.
+		routines: Set[Routine],
+
+		// Service implementation.
+		started: bool,
+		stopped: bool,
+
+		// Indicates whether a switch was set: reactor registration.
+		switch: bool,
+	}
+
+	var state: ReactorState
+
+	/*
+	 * GetChannels returns the list of MConnection.ChannelDescriptor. Make sure
+	 * that each ID is unique across all the reactors added to the switch.
+	 */
+	// TODO: we need a way to configure the returned channels!
+	def GetChannels(s: ReactorState): Set[ChannelDescriptor] = {
+		Set({ID: 3, Priority: 1})
+		.union(Set({ID: 7, Priority: 2}))
+		//.union(Set({ID: 3, Priority: 2})) // registration should fail!
+	}
+
+	/*
+	 * SetSwitch allows setting a switch.
+	 */
+	def SetSwitch(s: ReactorState, switch: bool): ReactorState = {
+		s.with("switch", switch)
+	}
+
+	// Code: Switch.AddReactor(name string, reactor Reactor)
+	action register() = all {
```

Comment:
I would separate pure defs and actions in the code.

Reply:
yeah. perhaps this is a matter of taste. Perhaps we can leave it like this for now. 
However, I think best readability would be
- pure defs
- vals (global state)
- actions (without guards)
- step (with guards to see correspondence to grammar)
but we can keep this issue for later.

Reply:
Yes, this is a good organization. Lets see what other reviewers think.

Reply:
Pure defs representing API methods and actions representing the p2p layer behavior are separated now.